aries_vcx/src/common/credentials/encoding.rs

```diff
@@ -1,99 +1,140 @@
+use std::collections::HashMap;
+
+use serde_json::Value;
+
 use crate::errors::error::{AriesVcxError, AriesVcxErrorKind, VcxResult};
 use crate::utils::openssl::encode;
-use std::collections::HashMap;
 
-pub fn encode_attributes(attributes: &str) -> VcxResult<String> {
-    let mut dictionary = HashMap::new();
-    match serde_json::from_str::<HashMap<String, serde_json::Value>>(attributes) {
-        Ok(mut attributes) => {
-            for (attr, attr_data) in attributes.iter_mut() {
-                let first_attr = match attr_data {
-                    // new style input such as {"address2":"101 Wilson Lane"}
-                    serde_json::Value::String(str_type) => str_type,
-
-                    // old style input such as {"address2":["101 Wilson Lane"]}
-                    serde_json::Value::Array(array_type) => {
-                        let attrib_value: &str = match array_type.get(0).and_then(serde_json::Value::as_str) {
-                            Some(x) => x,
-                            None => {
-                                return Err(AriesVcxError::from_msg(
-                                    AriesVcxErrorKind::InvalidAttributesStructure,
-                                    "Attribute value not found",
-                                ));
-                            }
-                        };
-
-                        warn!("Old attribute format detected. See vcx_issuer_create_credential api for additional information.");
-                        attrib_value
-                    }
-                    _ => {
-                        return Err(AriesVcxError::from_msg(
-                            AriesVcxErrorKind::InvalidJson,
-                            "Invalid Json for Attribute data",
-                        ));
-                    }
-                };
-
-                let encoded = encode(first_attr)?;
-                let attrib_values = json!({
-                    "raw": first_attr,
-                    "encoded": encoded
-                });
-
-                dictionary.insert(attr.to_string(), attrib_values);
+/// `CredentialAttributeData` represents contains all the attributes data
+/// in a credential.
+#[derive(Debug, Clone)]
+pub struct CredentialAttributeData<D> {
```

Comment:
This is currently as well pretty much just wrapping something what can be turned into `&str`, so lets either. I suggest you keep `encode` input signature as was, simply taking `&str`. 

Nevertheless, it would be actually amazing if the input argument for `encode_attributes` can be  dealt with properly. We can make that separate issue, separate PR if you'd like. Here's bit more detailed suggestion how to go about it:

The main issue is that `encode_attributes` supports 3 different input formats are. To make the `encode` function signature nicer (and implementation too) we need to fix the problem from its root. 
The 3 different formats currently supported:
```
{"address2":"101 Wilson Lane", "surname": "Smith"}
```
```
{"address2":["101 Wilson Lane"], "surname": ["Smith"]}
```
```
[{"name": "address2", "value": "101 Wilson Lane"}, {"name": "surname", "value": "Smith"}]
```
and apparently, the current implementation just does not care up until `encode` is called. Not only this makes it hard to create input type for `encode_attributes` function to accept, it also doesn't follow the "fail early" principle. 

So here's idea:
- find original place where this ambiguous string data is originating from - hint: it's when issuer is building credential offer msg `Issuer::build_credential_offer_msg`; `issuer.rs`) and parse the data there
- let's select the most sensible format and internally stick with that - probably `{"address2":"101 Wilson Lane", "surname": "Smith"}` makes sense the most. And therefore we can also type it less ambiguously than just `String` - we can now work with `HashMap<String, String>` instead.
- we can then therefore update `OfferSentState`, `RequestReceivedState` to work with that instead of `String`.
- the `encode_attributes` can then take `HashMap<String, String>` instead of `String` as well, and the implementation will be much simpler as we have compile time guarantee about the data we'll be processing.
Unless we implement some deserialization, this change will be breaking change for state machines in `OfferSentState`, `RequestReceivedState` states, but arguably it's low impact and acceptable.

Reply:
If we enumerate all the currently supported formats, we could limit the number of breaking changes in the state machines.

Reply:
That's true, but the implementation will be a lot simpler if we drop it, the impact won't be significant. Speaking for Absa, we don't mind and I am not aware of other parties currently using Issuer state machine (aries-vcx users seem to be typically on mobile side building native apps)

Quoting myself
> but arguably it's low impact and acceptable.
to elaborate on this, the state machines would rarely be long-term stored in `OfferSentState`, `RequestReceivedState ` long term. Usually entire credential exchange happens within a seconds.